modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityUpdate.C

```diff
@@ -0,0 +1,833 @@
+  }
+}
+
+// Calculates stress residual equation and jacobian
+void
+CrystalPlasticityUpdate::calculateResidualAndJacobian()
+{
+  calcResidual();
+  if (_error_tolerance)
+    return;
+  calcJacobian();
+}
+
+void
+CrystalPlasticityUpdate::calcResidual()
+{
+  RankTwoTensor ce, elastic_strain, ce_pk2, equivalent_slip_increment, pk2_new;
+
+  equivalent_slip_increment.zero();
+
+  for (unsigned int i = 0; i < _number_slip_systems; ++i)
+    _tau[_qp][i] = _pk2[_qp].doubleContraction(_flow_direction[_qp][i]);
+
+  // Call the overwritten method in the inheriting class that contains the constitutive model
+  calculateConstitutiveEquivalentSlipIncrement(equivalent_slip_increment, _error_tolerance);
+
+  if (_error_tolerance)
+    return;
+
+  RankTwoTensor residual_equivalent_slip_increment =
+      RankTwoTensor::Identity() - equivalent_slip_increment;
+  _inverse_plastic_deformation_grad =
+      _inverse_plastic_deformation_grad_old * residual_equivalent_slip_increment;
+
+  _elastic_deformation_gradient =
+      _temporary_deformation_gradient * _inverse_plastic_deformation_grad;
+
+  ce = _elastic_deformation_gradient.transpose() * _elastic_deformation_gradient;
+  elastic_strain = ce - RankTwoTensor::Identity();
+  elastic_strain *= 0.5;
+
+  pk2_new = _elasticity_tensor[_qp] * elastic_strain;
+  _residual_tensor = _pk2[_qp] - pk2_new;
+}
+
+void
+CrystalPlasticityUpdate::calcJacobian()
+{
+  RankFourTensor dfedfpinv, deedfe, dfpinvdpk2;
+
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+      for (unsigned int k = 0; k < LIBMESH_DIM; ++k)
+        dfedfpinv(i, j, k, j) = _temporary_deformation_gradient(i, k);
+
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+      for (unsigned int k = 0; k < LIBMESH_DIM; ++k)
+      {
+        deedfe(i, j, k, i) = deedfe(i, j, k, i) + _elastic_deformation_gradient(k, j) * 0.5;
+        deedfe(i, j, k, j) = deedfe(i, j, k, j) + _elastic_deformation_gradient(k, i) * 0.5;
+      }
+
+  calculateTotalPlasticDeformationGradientDerivative(dfpinvdpk2);
+
+  _jacobian =
+      RankFourTensor::IdentityFour() - (_elasticity_tensor[_qp] * deedfe * dfedfpinv * dfpinvdpk2);
+}
+
+void
+CrystalPlasticityUpdate::calculateTotalPlasticDeformationGradientDerivative(
+    RankFourTensor & dfpinvdpk2)
+{
+  calculateConstitutivePlasticDeformationGradientDerivative(
+      dfpinvdpk2, _flow_direction[_qp], _number_slip_systems);
+}
+
+void
+CrystalPlasticityUpdate::calculateConstitutivePlasticDeformationGradientDerivative(
+    RankFourTensor & dfpinvdpk2,
+    std::vector<RankTwoTensor> & schmid_tensor,
+    const unsigned int & number_dislocation_systems,
+    unsigned int /*slip_model_number*/)
+{
+  std::vector<Real> dslip_dtau(number_dislocation_systems, 0.0);
+  std::vector<RankTwoTensor> dtaudpk2(number_dislocation_systems);
+  std::vector<RankTwoTensor> dfpinvdslip(number_dislocation_systems);
```

Comment:
if `number_dislocation_systems` does not change and can be determined during construction, it will be better to have those be members.

Reply:
After a long trip down memory lane / finding the right notebook and git commit, it looks like I'd made this a local variable so that I could use these methods for different dislocation movement mechanisms (in my case glide and twinning), passing in the associated number of systems for both of those different mechanisms. Is there a good way for me to keep this flexibility while making this a member variable?

Reply:
I could switch this to the `_number_slip_systems` member variable for now, and leave this sort of extension for later work. Which would be most preferable?

Reply:
You could still make it a member and size it to the maximum value of all `number_dislocation_systems`. This could be done dynamically on demand with

```
if (_dfpinvdslip.size() < number_dislocation_systems)
  _dfpinvdslip.resize(number_dislocation_systems);
```

Reply:
I won't know what the maximum value is for `number_dislocation_system` at construction time: that's something that the inheriting classes would only know. 

e.g. pretend we have a child class with 12 glide slip systems (`_number_slip_systems`) and, let's say, 6 twin systems (`_number_twin_systems). This base class doesn't know `_number_twin_systems` exists. The way I intended this method call to work is that number of systems for the active mechanism are looped over in this calculation (this is probably obvious to you). In this made-up example, would the maximum value of `_number_dislocation_systems` be 18 or 12? And would the inheriting classes that add additional mechanisms be responsible for checking and updating the value of `_number_dislocation_systems` in their constructor?

Reply:
Alternatively, the passed in `schmid_tensor` already has the size that corresponds to that pesky `number_dislocation_systems`, so I can just query the size of that vector to set up the loop properly. That's my new favorite option; anything particularly horrible about that option?